### PR TITLE
New Example: Stacked Autoencoder

### PR DESCRIPTION
**Pull request by mthrok**
Demonstration of Autoencoder module and reuse of trained encoders.

### CODE REVIEW

File: examples/mnist_sae.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+from keras.datasets import mnist
+from keras.models import Sequential
+from keras.layers import containers
+from keras.layers.core import Dense, AutoEncoder
+from keras.utils import np_utils
+
+'''
+    Example of Stacked Autoencoder.
+    This is not an interesting example but to demonstrate how to use AutoEncoder
```
Comment (fchollet) [inline examples/mnist_sae.py]: > This is not an interesting example

Then what is its purpose?

**Comment by fchollet** [comment]
This example would probably answer some questions that many newcomers to DL seem to have about AEs and pretraining, but I have an issue with "endorsing" badly outdated techniques by including them in the examples folder. This is encouraging people to use methods they definitely should stay away from unless they really know what they are doing.

What are some arguments why we should include this?

**Comment by mthrok** [comment]
Well, if you think so while understanding my intent, then there is no argument on it.

**Comment by sahilzubair** [comment]
@fchollet I have been reading a number of comments on issues regarding autoencoders and their 'useleness'. A couple things I would like to point out - the only way I can imagine to reduce data using core layers in keras is with an autoencoder. They are also very useful for creating representations so that it is possible to visualize highly dimensional data. Maybe I will find time to contribute and show a graph of word embeddings that are dimensionally reduced and graphed by an auto encoder.

File: examples/mnist_sae.py (hunk 1 above)
Comment (belevtsoff) [inline examples/mnist_sae.py]: @fchollet I find this one very useful, why not merge it?

File: examples/mnist_sae.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    # Store trainined weight
+    trained_encoders.append(ae.layers[0].encoder)
+    # Update training data
+    X_train_tmp = ae.predict(X_train_tmp)
+
+# Fine-tuning
+print('Fine-tuning')
+model = Sequential()
+for encoder in trained_encoders:
+    model.add(encoder)
```
Comment (perone) [inline examples/mnist_sae.py]: When you call `add()` method here, wouldn't it call the `set_previous()` method and then reinitialize the weights again ?

Comment (bobobo1618) [inline examples/mnist_sae.py]: It appeared to do so. I altered the `append` line above to: `trained_encoders.append((ae.layers[0].encoder, ae.layers[0].encoder.get_weights()))` and this line to:

```
model.add(encoder)
model.layers[-1].set_weights(weights)
```

This seems to be working pretty well.

**Comment by weiliu620** [comment]
I would suggest to merge this example. AE itself may be outdated but there are some work based on AE, such as the ladder network in NIPS this year. So this example will be a good starting point for beginners. 

I haven't read the code, though.

**Comment by snurkabill** [comment]
+1 for example

**Comment by hanlianlu** [comment]
@fchollet  What if the most part of data I have is unlabelled,  is Pre-training in this situation with Autoencoder quite beneficial ?

**Comment by bajibabu** [comment]
This code works for the keras version == 0.3.2.
`
from __future__ import absolute_import
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras import models
from keras.layers import containers, AutoEncoder, Dense

np.random.seed(1337) # for reproducibility

def train_SDA(X_train, X_test, hidden_layers, do_pretrain=True, 
	hidden_activation='tanh', output_activation='linear',
	batch_size=256, pre_train_epoch=10, nb_epoch=25):
	
	#activation = 'tanh'
	# Layer-wise pre-training
	pre_trained_layers = []
	X_train_tmp = X_train
	for n_in, n_out in zip(hidden_layers[:-1], hidden_layers[1:]):
		print('Pre-training the layer: Input {} -> Output {}'.format(n_in, n_out))
		
		# Create AE and training 
		encoder = containers.Sequential([Dense(n_out, input_dim=n_in, 
			              activation=hidden_activation)])
		decoder = containers.Sequential([Dense(n_in, input_dim=n_out, 
			              activation=hidden_activation)])
		ae = AutoEncoder(encoder=encoder, decoder=decoder, 
			              output_reconstruction=True)
		ae_model = models.Sequential()
		ae_model.add(ae)
		
		# Train the simple autoencoder
		ae_model.compile(loss='mse', optimizer='rmsprop')
		ae_model.fit(X_train_tmp, X_train_tmp, batch_size=batch_size, 
			              nb_epoch=pre_train_epoch)
		
		# Store trainined weight
		pre_trained_layers.append((ae_model.layers[0].encoder, 
			              ae_model.layers[0].encoder.get_weights()))
		
		# Predict hidden nodes output
		ae.output_reconstruction = False
		ae_model.compile(loss='mse', optimizer='rmsprop')
		
		# Update training data
		X_train_tmp = ae_model.predict(X_train_tmp)

	# Fine-tuning
	print('Fine-tuning')
	# Create a new model
	model = models.Sequential()
	# Add the pre-trained layers to the model
	for layer, weights in pre_trained_layers:
	    model.add(layer)
	    model.layers[-1].set_weights(weights)

	# Add the output layer with linear activation function
	model.add(Dense(hidden_layers[0], input_dim=hidden_layers[-1], 
		activation=output_activation))

	# Train the model
	model.compile(loss='mse', optimizer='rmsprop')
	model.fit(X_train, X_train, batch_size=batch_size, nb_epoch=nb_epoch,
	          show_accuracy=True, validation_data=(X_test, X_test))

	score = model.evaluate(X_test, X_test, show_accuracy=True, verbose=0)
	print('Test score:', score[0])
	print('Test accuracy:', score[1])

if __name__ == '__main__':
	
	batch_size = 640
	nb_classes = 10
	nb_epoch = 3
	hidden_layers = [784, 600]

	# the data, shuffled and split between train and test sets
	(X_train, y_train), (X_test, y_test) = mnist.load_data()
	X_train = X_train.reshape(-1, 784)
	X_test = X_test.reshape(-1, 784)
	X_train = X_train.astype("float32") / 255.0
	X_test = X_test.astype("float32") / 255.0
	print(X_train.shape[0], 'train samples')
	print(X_test.shape[0], 'test samples')

	train_SDA(X_train, X_test, hidden_layers, do_pretrain=True, 
	hidden_activation='tanh', output_activation='linear',
	batch_size=256, pre_train_epoch=2, nb_epoch=2)
`

**Comment by ylqfp** [comment]
Great! This example helps me a lot!

**Comment by zeka0** [comment]
Hadn't anyone found that the codes provided by @bajibabu has a serious problem?
Look at it closely, it's actually a very very special case: **there is only one hidden layer**.
Because if _his code won't work if there are more than 2 hidden layers_.

Reasons are as follows:
If there are more than 1 hidden layers.
The training samples' shape won't match the second autoencoder input shape.

We need a way to tell autoencoders to use outputs of previous hidden layers.

**Comment by bajibabu** [comment]
Hello, I gave an example.. you can easily extend the above models to many hidden layers

**Comment by zeka0** [comment]
Yeah....I know. But I need to point that out, right?..

**Comment by xiaofangyuan** [comment]
I am not sure, but it seems like a question that  when we do this step
# Update training data
+    X_train_tmp = ae.predict(X_train_tmp)
The predict output dimension is not right, because it should use the encoder layer parameters to predict but not the autoencoder model?
